File: scripts/pruneFixedcM.py
# this script prunes a set of SNPs, read from a gzipped tab deliminated file,
# so that no two are closer than L cM (RECOMBINATION DISTANCE -- to reduce LD)

# to run: hilo/scripts/$ python3 pruneFixedcM.py 0.0001 55 0.1 ../data/TEST/10SNPS.prunedR.mafs ../data/TEST/10SNPS.mafs.gz ../data/TEST/10SNPS.B.mafs.gz
import sys
import csv
import gzip
import pandas
import calcMapPos # helper function to calculate map position in cM from bp pos


# get input 
minL = float(sys.argv[1]) # minimum length (in cM) between 2 SNPs that are kept 
minInd = int(sys.argv[2]) # minimum number of individuals with data
minMAF = float(sys.argv[3]) # column number specifying 
fileOut = sys.argv[4] # file/path to write results to (.gz will be added for gzip compressed)
listIn = sys.argv[5:] # list of files/paths to read in (gzipped) 

print("Pruning R: minL= " + str(minL) + "cM, minInd= " + str(minInd) + " minMAF= " + str(minMAF) + " fileOut= " + fileOut + " listIn= " + str(listIn))

# set values
# chromo	position	major	minor	phat	nInd
colChr = 0 # column number of SNP chromosome/scaffold/LG. (index starts at zero!)
colPosBP = 1 # column number of SNP position column (index starts at zero!)
colMAF = 4 # column number specifying minimum allele freq. of SNPs to include
colNInd = 5 # column number specifying number of individuals with data to include a SNP

# get linkage map
rmapALL = pandas.read_csv("../data/linkage_map/ogut_fifthcM_map_agpv4.txt", sep = "\t", header = None, names = ["name", "marker", "pos_cM", "chrom", "pos_bp"])

pos1 = None # starting position
chr1 = None # starting chromosome
# first output file has all included SNPs with original bp positions, and is gzipped
with gzip.open(fileOut + ".gz", mode = "wt") as fileSNP:
    writerSNP = csv.writer(fileSNP, delimiter = "\t")
    # second output file has recombination position only for all included SNPs
    with open(fileOut + ".rpos", mode = "w") as writerPOS:
        # for each input file, find SNPs to include
        for fileIn in listIn:
            with gzip.open(fileIn, mode = "rt") as fileRead:
                reader = csv.reader(fileRead, delimiter = "\t")
                for row in reader:
                    try:
                        posBP2 = int(row[colPosBP])
                    except ValueError: # if row[colPosBP] isn't an integer, skip and go to next line because current line is a header
                        print("skipping header line: " + str(row))
                        continue
                    if float(row[colMAF]) < minMAF or int(row[colNInd]) < minInd:
                        #print("no pass, MAF=" + row[colMAF] + ", NInd=" + row[colNInd])
                        pass # doesn't meet minimum criteria
                    else: # SNP passes minimum criteria
                        # get new chrom
                        chr2 = int(row[colChr])
                        # get new recombination position 
                        pos2 = calcMapPos.calcMapPos(chrom = chr2, pos = posBP2, rmap = rmapALL)
                        #print("at SNP " + str(chr2)+ ":" + str(pos2) + "cM") 
                        if chr2 == chr1 and (pos2 - pos1) < minL: # if same chromosome and too close, skip SNP
                            if pos1 is not None and pos2 < pos1:
                                raise ValueError("Positions must be in order on each chromosome! Error at Chr" + str(chr1) + ":" + str(pos1) + " coming before Chr" + str(chr2) + ":" + str(pos2) + " at bp-position " + str(posBP2) + " file " + fileIn)
                            pass 
                        else: # include if on diff chrom or not too close
                            writerSNP.writerow(row) # print line (SNP included)
                            if chr2 != chr1: # first SNP included from a chromosome
                                writerPOS.write(str(1) + "\n") # arbitrary but starts with difference = 1
                            else: # print difference in Morgans
                                writerPOS.write(str((pos2 - pos1)/100) + "\n") 
                            chr1 = chr2 # update current chromosome
                            pos1 = pos2 # update current position
print("Done pruning by fixed cM distance.")

File: scripts/pruneFixedcM.R
# THIS IS A DRAFT SCRIPT -- NOT NEEDED AT THE MOMENT BECAUSE OF 
# pruneFixedcM.py
# inputs (e.g. from last processed file & other variables for thinning):
input_kept_cM = 0
input_kept_chr = 0 # not a true chromosome
minDist = .0001 #(.001cM ~ 1Kb on average across the genome)

regions = region_start:region_end # takes in consecutive set of regions to apply thinning
region_start = 9
region_end = 9

# load recombination map and make a function to calculate map position of any SNP
rmap = read.table("../data/linkage_map/ogut_fifthcM_map_agpv4_INCLUDE.txt",
                  sep = "\t", header = F, stringsAsFactors = F)
colnames(rmap) = c("SNP", "marker", "pos_cM", "chr", "pos_bp")

# helper function to calculate map position (also implemented in calcMapPos.py)
calcMapPos = function(chrom, pos, rmap){
  # subset recombination map to only relevant chromosome
  rmapChr = rmap[rmap$chr == chrom, ]
  # if position is outside the range of the map 
  # (less than 1st value or greater than last value)
  # use avg. recombination rate for the closest bin (at that end of the chr)
  if (pos < rmapChr$pos_bp[1]){
    #print("off map left")
    left = rmapChr[1, ] # leftmost mapped position
    right = rmapChr[2, ] # next position from furthest left
  } else{
    if (pos >= rmapChr$pos_bp[nrow(rmapChr)]){
      left = rmapChr[nrow(rmapChr)-1, ] # second to last position from the right
      right = rmapChr[nrow(rmapChr), ] # rightmost mapped position on chromosome
    } else{ # otherwise if SNP is within the map, find the closest mapped positions above and below pos
      left = rmapChr[rmapChr["pos_bp"] <= pos, ][sum(rmapChr$pos_bp <= pos), ] # last mapped position smaller than or equal to pos
      right = rmapChr[rmapChr["pos_bp"] > pos, ][1, ] # first mapped position greater than pos
    }
  }
  # calculate cM/bp in local region between 2 closest mapped positions
  rate = (right$pos_cM - left$pos_cM) / (right$pos_bp - left$pos_bp)
  #print("cM/bp rate is " + str(rate)) 
  # use this rate to find distance from leftmost position
  # note that the bp difference * rate will be negative if the position is off the map to the left of all mapped positions
  map_pos = left$pos_cM + (pos - left$pos_bp)*rate 
  
  if (rate <= 0){ # should always be true with cleaned recombination rate map
    stop(paste("Recomb. rate calculated is NOT strictly positive, pos = ", pos))
  } 
  return(map_pos) # returns position in cM (note: may be negative! b/c mapped positions go negative from some previous 0)
}

# function to thin SNPs based on a minimum cM distance between SNPs
thin_SNPs = function(d, rmap, D, minInd, minDist, first_cM, first_chr){
  last_kept_cM = first_cM
  last_kept_chr = first_chr
  # now filter SNPs for LD
  # saves last kept SNP and a vector of T/F for which SNPs are to be kept
  keep = rep(F, nrow(d))
  d1 = d %>%
    mutate(., cM = sapply(1:nrow(.), function(i) 
      calcMapPos(chrom = chromo[i], pos = position[i], rmap = rmap)))
  for (i in 1:length(keep)){
    # always in low LD if on diff. chromosomes
    cM1 = d1[i, "cM"]
    chr1 = d1[i, "chromo"]
    keep[i] = abs(cM1 - last_kept_cM) >= minDist | chr1 != last_kept_chr
    if (keep[i]){ # if you keep SNP, update last-kept-SNP = current NSP
      last_kept_cM = cM1
      last_kept_chr = chr1
    }
  }
  d_thin = d1[keep, ]
  # calculate diference in morgans between positions kept
  d_thin$diff_M = diff(c(first_cM, d_thin$cM))/100
  ifelse(first_chr != last_kept_chr){ # if this file is the first of a new chromosome
    d_thin$diff_M[1] = 1 # make Morgan difference = 1 (arbitrary for ancestry_hmm input)
  }
  return(d_thin)
}


# print out a file with chromosome, position, major, minor & difference between adjacent 
# SNPs in Morgans (to 6 or 8 decimal places .. I'd have to check)
# I need to ultimately get 1 file per chromosome after thinning .. but I could concatenate after thinning
# or I could still concatenate at the very end after getting frequencies etc.
last_cM = 0
last_chr = 0
for (r in regions){
  
  d = thin_SNPs(d = getSNPs(region = r, 
                            dir_gl = dir_gl, 
                            dir_sites = dir_sites, 
                            allo_maize = allo_maize, 
                            allo_mex = allo_mex), 
                rmap = rmap, 
                D = D, 
                minInd = minInd, 
                minDist = minDist, 
                first_cM = last_cM, 
                first_chr = last_chr)
  # write to file
  write.table(paste0(dir_sites, "/thinned/")) ######### NOT FINISHED
}


